torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -119,6 +128,78 @@ def main_hook(self):
         self.zero.step()
 
 
+class _DDPBucket(NamedTuple):
+    r"""
+    This contains the model parameters corresponding to a
+    :class:`DistributedDataParallel` gradient bucket.
+
+        bucket_index (int): index of the bucket determined by the DDP gradient
+            bucket all-reduce order.
+        params (List[torch.Tensor]): model parameters in the bucket.
+    """
+    bucket_index: int
+    params: List[torch.Tensor]
+
+
+class _OverlapStatus(enum.IntEnum):
+    r"""
+    This defines the three possible statuses that
+    :class:`ZeroRedundancyOptimizer` can be in when overlapping with
+    :class:`DistributedDataParallel`.
+
+        ``UNINITIALIZED``: The ZeRO instance is effectively uninitialized and
+            is waiting for DDP to finalize its bucketing.
+        ``DDP_HAS_REBUILT_BUCKETS``: DDP has rebuilt its buckets, meaning that
+            its bucketing is finalized. The ZeRO instance can now collect the
+            necessary information about the DDP bucketing.
+        ``INITIALIZED``: The ZeRO instance is fully initialized and can now
+            optimize parameters.
+    """
+    UNINITIALIZED = 0
+    DDP_HAS_REBUILT_BUCKETS = 1
+    INITIALIZED = 2
+
+
+class _OverlapInfo():
+    r"""
+    This contains the information needed by :class:`ZeroRedundancyOptimizer`
+    to overlap with :class:`DistributedDataParallel`.
+
+        status (_OverlapStatus): current status; see :class:`_OverlapStatus`
+            for more information.
+        params_per_bucket (List[List[torch.Tensor]]): ``params_per_bucket[i]``
+            gives the model parameters in the ``i``th bucket.
+        params_per_rank (List[List[torch.Tensor]]): ``params_per_rank[i]``
+            gives the model parameters assigned to the ``i``th rank, where the
+            parameters are grouped by increasing bucket indices.
+        offsets (Dict[int, int]): maps from bucket index to the offset in
+            ``self.params_per_rank[rank]`` giving the index of the first
+            parameter in that bucket, where ``rank`` is this process's own
+            rank; the keys of this :class:`dict` are the bucket indices
+            assigned to this rank.
+        broadcast_handles (List[Work]): :class:`list` of async work handles for
+            the parameter broadcasts.
+        bucket_index_to_future (Dict[int, torch.futures.Future]):
+            :class:`dict` mapping bucket index to the corresponding all-reduce
+            future.
+        bucket_index_to_bucket (Dict[int, dist.GradBucket]): :class:`dict`
```

Comment:
nit (can be done in follow PRs): curious, callbacks on buckets will fire in order, so this can be simplified to a list of GradBucket objects?

Reply:
For a given rank, we only need to store references to the `GradBucket`s that the rank uses to update. As such, even though the buckets fire in order, the "keys" for this data structure are only a subset of all indices. 

If the total number of buckets is not large, using a `list` and either storing all buckets or storing `None`s in unused positions may be faster than storing a `dict`, but I used a `dict` for now.

Reply:
Ah, I see, makes sense.